### PR TITLE
Add configs for transaction log properties and allow disabling Atomikos transaction manager

### LINKED ISSUE #26053: Add atomikos file persistence log location
**Description:**
The log location should be read from the config-reader and loaded to atomikos

### PR DESCRIPTION
**Pull request by lasinicl**
Fixes #26053, #26055

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 1]
```diff
@@ -160,7 +235,33 @@ public void registerParticipation(String gTransactionId, String transactionBlock
...
+            return true;
+        }
+        String combinedId = generateCombinedTransactionId(transactionId, transactionBlockId);
+        List<BallerinaTransactionContext> txContextList = resourceRegistry.get(combinedId);
+        if (txContextList != null) {
+            for (BallerinaTransactionContext ctx : txContextList) {
+                try {
+                    XAResource xaResource = ctx.getXAResource();
+                    if (xaResource != null) {
+                        Xid xid = xidRegistry.get(combinedId);
```
Comment (mohanvive) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:248]: Can't we move this line out of `for` loop?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 2]
```diff
@@ -160,7 +235,33 @@ public void registerParticipation(String gTransactionId, String transactionBlock
...
+                        xaResource.prepare(xid);
+                    }
+                } catch (Throwable e) {
+                    log.error("error in prepare the transaction, " + combinedId + ":" + e.getMessage(), e);
+                    return false;
```
Comment (mohanvive) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:253]: Why don't we throw this exception? Any reason to handle it and return `falase`?

Comment (mohanvive) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:253]: Again, error message does not looks good. Shall we change it as "Error at transaction prepare phase" ?

Comment (mohanvive) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:253]: Also, whether printing combinedid makes sense? Don't we need to print transaction id and transaction block id separately?

**Review by mohanvive — CHANGES_REQUESTED** [review]

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 2 above)
Comment (mohanvive) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:254]: @lasinicl Need this change in other places as well.

Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:253]: Fixed

Comment (gimantha) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:253]: TransactionBlockId is something internal right? it is ok to print it in error log? May be ok as a debug log.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 3]
```diff
@@ -96,6 +119,57 @@ public static TransactionResourceManager getInstance() {
         return transactionResourceManager;
     }
 
+    /**
+     * This method adds configs for transaction log path and transaction log name.
+     *
+     */
+    private void addLogConfigToRegistry() {
```
Comment (gimantha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:126]: shall we add a test case for this in transaction repo as a unit test?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 4]
```diff
@@ -96,6 +119,57 @@ public static TransactionResourceManager getInstance() {
...
+    }
+
+    /**
+     * This method check whether the atomikos transaction manager should be enabled or not.
+     *
+     * @return boolean whether the atomikos transaction manager should be enabled or not
+     */
+    private boolean getTransactionManagerEnabled() {
+        String transactionManagerEnabled = CONFIG_REGISTRY.getAsString("b7a.transaction.manager.enabled");
+        if (transactionManagerEnabled != null) {
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:165]: merge the ifs

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 5]
```diff
@@ -96,6 +119,57 @@ public static TransactionResourceManager getInstance() {
...
+    /**
+     * This method adds configs for transaction log path and transaction log name.
+     *
+     */
+    private void addLogConfigToRegistry() {
+        // Path to log directory given for atomikos log path property should exist.
+        // Therefore create a transaction_log_dir in root directory if not exists.
+        final Path projectRoot = findProjectRoot(Paths.get(System.getProperty("user.dir")));
+        if (projectRoot != null) {
+            String logPath = projectRoot.toAbsolutePath().toString() + "/transaction_log_dir";
```
Comment (gimantha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:131]: Is it possible for your to define the log directory as well? if he as not set it, we will use this name as the default one.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 2 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:241]: Fixed

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 4 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java]: fixed

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 5 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java]: fixed

**Review by gimantha — APPROVED** [review]

**Review by gimantha — APPROVED** [review]

**Review by mohanvive — APPROVED** [review]

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 6]
```diff
@@ -96,6 +118,51 @@ public static TransactionResourceManager getInstance() {
...
 
+    /**
+     * This method sets values for atomikos transaction log path and name properties using the available configs.
+     *
+     */
+    private void setLogProperties() {
+        final Path projectRoot = findProjectRoot(Paths.get(System.getProperty("user.dir")));
+        if (projectRoot != null) {
+            String logDir = getTransactionLogDirectory();
+            String logPath = projectRoot.toAbsolutePath().toString() + "/" + logDir;
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:129]: Use File.separatorChar for platform independent coding while joining paths

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 7]
```diff
@@ -96,6 +118,51 @@ public static TransactionResourceManager getInstance() {
...
+            Path transactionLogDirectory = Paths.get(logPath);
+            if (!Files.exists(transactionLogDirectory)) {
+                try {
+                    Files.createDirectory(transactionLogDirectory);
+                } catch (IOException e) {
+                    stderr.println("error: failed to create transaction log directory");
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:135]: Improve the error msg to include the directory

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 8]
```diff
@@ -96,6 +118,51 @@ public static TransactionResourceManager getInstance() {
...
+        }
+    }
+
+    /**
+     * This method checks whether the atomikos transaction manager should be enabled or not.
+     *
+     * @return boolean whether the atomikos transaction manager should be enabled or not
+     */
+    private boolean getTransactionManagerEnabled() {
+        boolean transactionManagerEnabled = CONFIG_REGISTRY.getAsBoolean("b7a.transaction.manager.enabled");
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:149]: Let's move b7a.transaction.manager.enabled to a constant

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 9]
```diff
@@ -96,6 +118,51 @@ public static TransactionResourceManager getInstance() {
...
+        return transactionManagerEnabled;
+    }
+
+    /**
+     * This method gets the user specified config for log directory name.
+     *
+     * @return string log directory name
+     */
+    private String getTransactionLogDirectory() {
+        String transactionLogDirectory = CONFIG_REGISTRY.getAsString("b7a.transaction.log.base");
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:159]: Here as well

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 10]
```diff
@@ -171,25 +265,36 @@ public boolean notifyCommit(String transactionId, String transactionBlockId) {
...
-                log.error("error when committing the transaction, " + combinedId + ":" + e.getMessage(), e);
-                commitSuccess = false;
             }
 
             for (BallerinaTransactionContext ctx : txContextList) {
                 try {
                     XAResource xaResource = ctx.getXAResource();
-                    if (xaResource == null) {
-                        ctx.commit();
+                    if (transactionManagerEnabled) {
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:284]: merge the two if stmnts

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 11]
```diff
@@ -219,25 +324,35 @@ public boolean notifyAbort(String transactionId, String transactionBlockId, Obje
...
-                log.error("error when aborting the transaction, " + combinedId + ":" + e.getMessage(), e);
-                abortSuccess = false;
             }
 
             for (BallerinaTransactionContext ctx : txContextList) {
                 try {
                     XAResource xaResource = ctx.getXAResource();
-                    if (xaResource == null) {
-                        ctx.rollback();
+                    if (transactionManagerEnabled) {
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:342]: merge these ifs as well

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 12]
```diff
@@ -359,19 +486,38 @@ public TransactionLocalContext getCurrentTransactionContext() {
...
-        Transaction trx = trxRegistry.get(combinedId);
-        if (trx != null) {
-            List<BallerinaTransactionContext> txContextList = resourceRegistry.get(combinedId);
-            if (txContextList != null) {
-                for (BallerinaTransactionContext ctx : txContextList) {
-                    try {
-                        XAResource xaResource = ctx.getXAResource();
-                        if (xaResource != null) {
-                            trx.delistResource(xaResource, TMSUCCESS);
+        if (transactionManagerEnabled) {
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:489]: get the trx out and merge these ifs as well

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java [hunk 13]
```diff
@@ -359,19 +486,38 @@ public TransactionLocalContext getCurrentTransactionContext() {
...
+                        } catch (IllegalStateException | SystemException e) {
+                            log.error("error in ending the XA transaction " + transactionId
+                                    + ":" + e.getMessage(), e);
+                        }
+                    }
+                }
+            }
+        } else {
+            Xid xid = xidRegistry.get(combinedId);
+            if (xid != null) {
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:509]: merge xid and txContextList != null checks

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 6 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:129]: Fixed in 91ca624

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 7 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:135]: Fixed in 91ca624

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 8 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:149]: Fixed in 91ca624

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 9 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:159]: Fixed in 91ca624

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 10 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:284]: Fixed in 91ca624

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 13 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:509]: Fixed in 91ca624

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java (hunk 11 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/TransactionResourceManager.java:342]: Fixed in91ca624

**Pull request merged**